### ISSUE TITLE
Clean root directory from source files

### ISSUE THREAD

**Issue by regevran** [scylladb/scylladb#20981]
Refs: #2695 

Move all (or most) source files (`hh` and `cc`) into dedicated sub directories.
The main task is to clean the root directory from sources, leaving it clean for build, config/install and docs.
When this task is completed, the number of files in the root of the project will be reduced from ~160 to ~35.

Note: the target directories the source files are moved to may be later changed by other activities under #2695. In other words, moving the files not necessarily increase the order in other directories other than the root directory.

**Comment by avikivity** [comment]
What's the idea? To create a src/ directory?

Doesn't that just move the problem somewhere else?

**Comment by regevran** [comment]
> To create a src/ directory?

Not at all.
The idea is to move the files to  _existing_ directories.
It also requires:
- changing existing build to include the move files
- change the `#include`s to the new location
- anywhere the moved code is used - add dependency in moved-to library if such does not exist yet.

**Comment by nyh** [comment]
I agree with @regevran's idea, but have no idea how difficult or ugly it might be to do it. We can just start doing it one file at at a time...

As an example, I looked at validation.hh, validation.cc.
It has three functions, is_cql_key_invalid(), validate_column_family(), validate_keyspace().

The first function is used only in db/commitlog/commitlog_replayer.cc, the second one used only in cql3/statements/*.cc (different files there), and the third only in api/*.cc (several files).

So it seems to me that each of these functions should be moved to a different directory... They don't really have anything in common!

**Comment by regevran** [comment]
I have actually started doing this work - it progresses in an extremely slow pace.
But yes, the idea is to do it one (hh and cc) file at a time, and do the best to move it to a correct place. However, remembering that this is just a phase and a later movement is allowed if required. 
Another issue to consider, because of the slow pace - do we want to allow small structural PRs or we want to wait for a big change?
I prefer small PRs but it increases the load of the CI. It is preferable because then we can create issue-per-file and assign them as a background tasks to many people. 

Please advise.

**Comment by nyh** [comment]
> I have actually started doing this work - it progresses in an extremely slow pace. But yes, the idea is to do it one (hh and cc) file at a time, and do the best to move it to a correct place.

I recommend that you do this change one or few files at a time - and don't wait for one mega-patch involving 150 files. Such a patch would be a real pain to review, and you'll need to rebase it constantly as people continue to change those top-level files.

> However, remembering that this is just a phase and a later movement is allowed if required. Another issue to consider, because of the slow pace - do we want to allow small structural PRs or we want to wait for a big change? I prefer small PRs but it increases the load of the CI. It is preferable because then we can create issue-per-file and assign them as a background tasks to many people.

I prefer the smaller patches. Of course, pick a reasonable size - if validation.hh and validation.cc come together, change them in one patch and not two PRs. If query-result-*.hh are all related (I don't know if they are, I just ran "ls" and noticed that common prefix :-)), then you can move them in one patch. I think the deciding factor is something you can reasonable explain the motivation and the change in one commit message.

Another benefit of multiple separate pull requests is that some can go in while others are being resisted. For example, perhaps you can get rid of validation.cc/validation.hh because it's obvious where they need to go, but it's harder to get rid of query-request.hh (I'm just picking a file by random) because there will be heated debates on where it needs to go. If you put all of them in one huge pull request, it will never go in.

**Comment by regevran** [comment]
Thanks @nyh - I will act accordingly.
To make it happen I am creating many-per-file Github issues, enabling all developers to participate. 

As for:
> there will be heated debates on where it needs to go

This is exactly the reason it is declared as multi-phases and allows multi-movement - to prevent endless debates.

**Comment by nyh** [comment]
> Thanks @nyh - I will act accordingly.
> To make it happen I am creating many-per-file Github issues, enabling all developers to participate. 

Please don't... There is not need to spam the issue list for people to participate, and no reason to presume at which resolution people will want to fix this (e.g, maybe for one set of 7 files it makes sense to move together..)

**Comment by avikivity** [comment]
> > Thanks [@nyh](https://github.com/nyh) - I will act accordingly.
> > To make it happen I am creating many-per-file Github issues, enabling all developers to participate.
> 
> Please don't... There is not need to spam the issue list for people to participate, and no reason to presume at which resolution people will want to fix this (e.g, maybe for one set of 7 files it makes sense to move together..)

Agree with @nyh

**Comment by regevran** [comment]
**TL;DR**
`./types` - CQL types
`./utils` - any general use type.

**`./utils` and `./types`**
Originally I thought that classes and structs that represent a data in a way that one treat it as (almost) a built-in type should be in `/types`.
Then I was convinced that these actually belong to `./utils`.
`./types` is for CQL types.
Since cleaning and modularization the code is a long and continues process, I am ok with this decision at the moment.
_**Later**_ we can do more things like: naming `./types` --> `./cql-types`, moving `./types` --> `./cql/types`, splitting `./utils` --> `./utils/types` and `./utils/functions` or any other idea we have in mind.
